parrot/src/pfs_service_cvmfs.cc

```diff
@@ -1165,82 +1169,73 @@ class pfs_service_cvmfs:public pfs_service {
 			return 0;
 		}
 
-		/* cvmfs_open does not work with directories (it gives a 'fail to fetch' error). */
-		if(S_ISDIR(d.mode)) {
-			errno = EISDIR;
+		if((!S_ISDIR(d.mode)) && (flags&O_DIRECTORY)) {
+			errno = ENOTDIR;
 			return 0;
 		}
 
 		debug(D_CVMFS,"open(%s)",name->rest);
-		int fd = compat_cvmfs_open(name->rest);
 
-		if(fd<0) return 0;
+		if(S_ISDIR(d.mode)) {
+			struct cvmfs_dirent d;
+
+			/*
+			If the root of the CVFMS filesystem is requested,
+			we must generate it internally, containing the
+			list of the known filesystems.
+			*/
+
+			if(!name->host[0]) {
+				pfs_dir *dir = new pfs_dir(name);
+				dir->append(".");
+				dir->append("..");
+				CvmfsFilesystemList::const_iterator i    = cvmfs_filesystem_list.begin();
+				CvmfsFilesystemList::const_iterator iend = cvmfs_filesystem_list.end();
+				for(; i != iend; ++i) {
+					cvmfs_filesystem *f = *i;
+					/* If the host begins with dot, then it is a wildcard entry. */
+					/* Otherwise, it is a normal entry. */
+					const char *host = f->host.c_str();
+					if(host && host[0]!='.') {
+						dir->append(host);
+					}
+				}
+				return dir;
+			}
 
-		return new pfs_file_cvmfs(name, cvmfs_active_filesystem, fd, d);
-	}
+			/*
+			Otherwise, go to CVMFS for the directory liting.
+			*/
 
-	pfs_dir *getdir(pfs_name * name) {
-		struct cvmfs_dirent d;
+			pfs_dir *dir = new pfs_dir(name);
 
-		/*
-		If the root of the CVFMS filesystem is requested,
-		we must generate it internally, containing the
-		list of the known filesystems.
-		*/
+			char **buf = NULL;
+			size_t buflen = 0;
 
-		if(!name->host[0]) {
-			pfs_dir *dir = new pfs_dir(name);
-			dir->append(".");
-			dir->append("..");
-			CvmfsFilesystemList::const_iterator i    = cvmfs_filesystem_list.begin();
-			CvmfsFilesystemList::const_iterator iend = cvmfs_filesystem_list.end();
-			for(; i != iend; ++i) {
-				cvmfs_filesystem *f = *i;
-				/* If the host begins with dot, then it is a wildcard entry. */
-				/* Otherwise, it is a normal entry. */
-				const char *host = f->host.c_str();
-				if(host && host[0]!='.') {
-					dir->append(host);
-				}
-			}
-			return dir;
-		}
+			int rc = compat_cvmfs_listdir(name->rest, &buf, &buflen);
 
-		/*
-		Otherwise, go to CVMFS for the directory liting.
-		*/
+			if(rc<0) return 0;
```

Comment:
Put `pfs_dir *dir = new pfs_dir(name);` below this if statement so we don't leak memory when `compar_cvmfs_listdir` fails.